src/vehicle_gui.cpp

```diff
@@ -223,6 +226,118 @@ void BaseVehicleListWindow::BuildVehicleList()
 	this->vscroll->SetCount(static_cast<int>(this->vehgroups.size()));
 }
 
+
+/** Cargo filter functions */
+/**
+ * Check whether a vehicle can carry a specific cargo.
+ * @param v The vehicle to be checked
+ * @param cid The cargo what we are looking for
+ * @return Whether the vehicle can carry the specified cargo or not
+ */
+static bool CDECL CargoFilter(const GUIVehicleGroup *vehgroup, const CargoID cid)
+{
+	const Vehicle *v = (*vehgroup).GetSingleVehicle();
+	if (cid == BaseVehicleListWindow::CF_ANY) {
+		return true;
+	} else if (cid == BaseVehicleListWindow::CF_NONE) {
+		for (const Vehicle *w = v; w != nullptr; w = w->Next()) {
+			if (w->cargo_cap > 0) {
+				return false;
+			}
+		}
+		return true;
+	} else if (cid == BaseVehicleListWindow::CF_FREIGHT) {
+		bool have_capacity = false;
+		for (const Vehicle *w = v; w != nullptr; w = w->Next()) {
+			if (w->cargo_cap > 0) {
+				if (IsCargoInClass(w->cargo_type, CC_PASSENGERS)) {
+					return false;
+				} else {
+					have_capacity = true;
+				}
+			}
+		}
+		return have_capacity;
+	} else {
+		for (const Vehicle *w = v; w != nullptr; w = w->Next()) {
+			if (w->cargo_cap > 0 && w->cargo_type == cid) {
+				return true;
+			}
+		}
+		return false;
+	}
+}
+
+static GUIVehicleGroupList::FilterFunction * const _filter_funcs[] = {
+	&CargoFilter,
+};
+
+/**
+ * Set cargo filter list item index.
+ * @param index The index to be set
+ */
+void BaseVehicleListWindow::SetCargoFilterIndex(byte index)
+{
+	if (this->cargo_filter_criteria != index) {
+		this->cargo_filter_criteria = index;
+		/* Deactivate filter if criteria is 'Show All', activate it otherwise. */
+		this->vehgroups.SetFilterState(this->cargo_filter[this->cargo_filter_criteria] != CF_ANY);
+		this->vehgroups.SetFilterType(0);
+		this->vehgroups.ForceRebuild();
+	}
+}
+
+/**
+ *Populate the filter list and set the cargo filter criteria.
+ */
+void BaseVehicleListWindow::SetCargoFilterArray()
+{
+	byte filter_items = 0;
+
+	/* Add item for disabling filtering. */
+	this->cargo_filter[filter_items] = CF_ANY;
+	this->cargo_filter_texts[filter_items] = STR_PURCHASE_INFO_ALL_TYPES;
```

Comment:
Here you're reusing a string from elsewhere, and again on L310, `STR_NONE`. It might be better to define new strings for both of these alongside `STR_CARGO_TYPE_FREIGHT`.

Reply:
I defined new strings.